scanpipe/pipes/d2d.py

```diff
@@ -387,3 +440,189 @@ def purldb_match(project, extensions, logger=None):
+                        from_resource=from_resource,
+                        to_resource=match,
+                        map_type=map_type,
+                        extra_data={
+                            "path_score": f"{len(current_parts)}/{path_parts_len-1}",
+                        },
+                    )
+
+        if any_map:
+            break
+
+
+def _is_minified_and_map_compiled_from_source(
+    to_resources, from_source, minified_extension
+):
+    """Return True if a minified file and its map were compiled from a source file."""
+    if not minified_extension:
+        return False
+
+    path = Path(from_source.path.lstrip("/"))
+    basename, extension = _get_basename_and_extension(path.name)
+
+    source_file_name = path.name
+    source_mapping = f"sourceMappingURL={basename}{minified_extension}.map"
+
+    minified_file = None
+    minified_map_file = None
+    for resource in to_resources:
+        if resource.path.endswith(minified_extension):
+            minified_file = resource
+        elif resource.path.endswith(f"{minified_extension}.map"):
+            minified_map_file = resource
+        if minified_file and minified_map_file:
+            if _source_mapping_in_minified(minified_file, source_mapping) and (
+                _source_content_in_map(minified_map_file, from_source)
+                or _source_in_map(minified_map_file, source_file_name)
+            ):
+                return True
+            break
+
+    return False
+
+
+def _source_mapping_in_minified(resource, source_mapping):
+    """Return True if a string contains a specific string in its last 5 lines."""
+    lines = resource.file_content.split("\n")
+    total_lines = len(lines)
+    # Get the last 5 lines.
+    tail = 5 if total_lines > 5 else total_lines
+    return any(source_mapping in line for line in reversed(lines[-tail:]))
+
+
+def _source_in_map(map_file, source_name):
+    """
+    Return True if the given source file name exists in the sources list of the
+    specified map file.
+    """
+    try:
+        with open(map_file.location) as f:
+            data = json.load(f)
+            sources = data.get("sources", [])
+            return any(source.endswith(source_name) for source in sources)
+    except json.JSONDecodeError:
+        return False
+
+
+def _sha1(content):
+    """Calculate the SHA-1 hash of a string."""
+    hash_object = hashlib.sha1(content.encode())
+    return hash_object.hexdigest()
+
+
+def _source_content_in_map(map_file, source_file):
+    """Return True if the given source content is in specified map file."""
+    origin_sha1 = source_file.sha1
+    try:
+        with open(map_file.location) as f:
+            data = json.load(f)
+            contents = data.get("sourcesContent", [])
+            return any(origin_sha1 == _sha1(content) for content in contents)
+    except json.JSONDecodeError:
+        return False
+
+
+def _get_basename_and_extension(filename):
```

Comment:
This seems very tight to the JS world, the function name, docstring and variables are too generic and not reflecting this.